File: node/04_pii_redact.mjs
// Redact PII from documents — 14 types detected with context-aware AI

import { readFileSync, writeFileSync } from "fs";

const API_KEY = process.env.DEEPREAD_API_KEY;
const BASE = "https://api.deepread.tech";
const filePath = process.argv[2];

if (!API_KEY) { console.error("Set DEEPREAD_API_KEY"); process.exit(1); }
if (!filePath) { console.error("Usage: node 04_pii_redact.mjs <document.pdf>"); process.exit(1); }

const form = new FormData();
form.append("file", new Blob([readFileSync(filePath)]), filePath.split("/").pop());

const { id: jobId } = await fetch(`${BASE}/v1/pii/redact`, {
  method: "POST",
  headers: { "X-API-Key": API_KEY },
  body: form,
}).then(r => r.json());

console.log(`Submitted: ${jobId}`);

let delay = 3000;
while (true) {
  await new Promise(r => setTimeout(r, delay));
  const result = await fetch(`${BASE}/v1/pii/${jobId}`, {
    headers: { "X-API-Key": API_KEY },
  }).then(r => r.json());

  console.log(`  Status: ${result.status}`);

  if (result.status === "completed") {
    const report = result.report || {};
    console.log("\n--- PII Detection Report ---");
    console.log(`Pages scanned: ${report.page_count}`);
    console.log(`Total redactions: ${report.total_redactions}`);
    for (const [type, info] of Object.entries(report.pii_detected || {})) {
      console.log(`  ${type}: ${info.count} found on pages ${info.pages} (avg: ${info.confidence_avg.toFixed(2)})`);
    }
    if (result.redacted_file_url) {
      const pdf = await fetch(result.redacted_file_url).then(r => r.arrayBuffer());
      const outPath = filePath.replace(/\.(\w+)$/, "_redacted.$1");
      writeFileSync(outPath, Buffer.from(pdf));
      console.log(`\nRedacted file saved: ${outPath}`);
    }
    break;
  } else if (result.status === "failed") {
    console.error(`Failed: ${result.error}`);
    break;
  }
  delay = Math.min(delay * 1.5, 15000);
}

File: python/04_pii_redact.py
"""Redact PII from documents — names, SSNs, credit cards, and 11 other types."""

import os
import sys
import json
import time
import requests

API_KEY = os.environ.get("DEEPREAD_API_KEY")
BASE = "https://api.deepread.tech"

if not API_KEY:
    print("Set DEEPREAD_API_KEY in your environment or .env file")
    sys.exit(1)

file_path = sys.argv[1] if len(sys.argv) > 1 else None
if not file_path:
    print("Usage: python 04_pii_redact.py <document.pdf>")
    sys.exit(1)

headers = {"X-API-Key": API_KEY}

# Submit for PII redaction
with open(file_path, "rb") as f:
    resp = requests.post(
        f"{BASE}/v1/pii/redact",
        headers=headers,
        files={"file": f},
    )
    resp.raise_for_status()

job = resp.json()
job_id = job["id"]
print(f"Submitted: {job_id}")

# Poll with backoff
delay = 3
while True:
    time.sleep(delay)
    result = requests.get(f"{BASE}/v1/pii/{job_id}", headers=headers).json()
    status = result["status"]
    print(f"  Status: {status}")

    if status == "completed":
        report = result.get("report", {})
        print(f"\n--- PII Detection Report ---")
        print(f"Pages scanned: {report.get('page_count')}")
        print(f"Total redactions: {report.get('total_redactions')}")
        print(f"Confidence threshold: {report.get('confidence_threshold_used', 0.85)}")
        print(f"Below threshold: {report.get('below_threshold_count', 0)}")

        for pii_type, info in report.get("pii_detected", {}).items():
            print(f"  {pii_type}: {info['count']} found on pages {info['pages']} (avg: {info['confidence_avg']:.2f}, min: {info.get('confidence_min', 0):.2f})")

        # Download redacted file
        redacted_url = result.get("redacted_file_url")
        if redacted_url:
            pdf = requests.get(redacted_url).content
            out_path = file_path.rsplit(".", 1)[0] + "_redacted." + file_path.rsplit(".", 1)[1]
            with open(out_path, "wb") as f:
                f.write(pdf)
            print(f"\nRedacted file saved: {out_path}")
        break
    elif status == "failed":
        print(f"Failed: {result.get('error', 'Unknown error')}")
        break

    delay = min(delay * 1.5, 15)
